src/tactic/lift.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2019 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import tactic.rcases
+
+/-!
+# lift tactic
+
+This file defines the lift tactic, allowing the user to lift elements from one type to another
+under a specified condition.
+
+## Tags
+
+lift, tactic
+-/
+
+universe variables u v
+
+run_cmd mk_simp_attr `can_lift
+
+/-- A class specifying that you can lift elements from `α` to `β` assuming `cond` is true.
+  Used by the tactic `lift`. -/
+class can_lift (α : Type u) (β : Type v) : Type (max u v) :=
+(coe : β → α)
+(cond : α → Prop)
+(prf : ∀(x : α), cond x → ∃(y : β), coe y = x)
+
+@[can_lift] instance : can_lift ℤ ℕ :=
+⟨coe, λ n, 0 ≤ n, λ n hn, ⟨n.nat_abs, int.nat_abs_of_nonneg hn⟩⟩
+
+namespace tactic
+
+meta def lift (p : pexpr) (t : pexpr) (h : option pexpr) (n : list name) : tactic unit :=
```

Comment:
I find the behavior of this tactic a little hard to follow. Are there reasonable subroutines that you could factor out? Maybe lines 45-55, for instance? I can imagine this being tricky to re-parse if we need to update it down the line. At least a few more comments would help.

Reply:
Better now?

Reply:
Much cleaner, thanks! `lift` needs its own doc string. (Since the main interface is the interactive version, the doc string should primarily just say what the arguments are.)

Reply:
done